File: datasets/Part 2 - Regression/Section 8 - Decision Tree Regression/decission_tree_regression.R
# Árbol de Decisión para Regresión

# CART (Classification and Regression Trees)
# =======================================================================================================
# 
#  Si graficamos en 2D datos de 2 variables independientes (X1,X2) que permitiran predecir  una variable
# dependiente, el algoritmo dividira los conjuntos de puntos con lineas rectas. El algoritmo mira la entropia,
# que tan juntos o dispersos pueden estar esos puntos o que similitudes tienen entre si. 
# Cada uno de estos conjuntos corresponde a un nodo hoja. Se continua haciendo divisiones hasta cierto
# punto, p.e. Cuando un nodo hoja se quede por lo menos al 5% de datos original.
# 
# La idea del algoritmo es que la cantidad de informacion aumenta (es mas acertada) cuando dividimos los 
# puntos en conjuntos o alguna otra regla, para hacer que el algoritmo converja y no haya Overffiting.
# 
# Al mismo tiempo que se divide el conjunto, se "genera" un arbol. si la primera division fuera que X1 < 20
# se generarian entonces 2 ramas. Luego si la sig. div. es que X2 < 170, pero solo en quellos con X1 > 20, 
# se generan 2 nuevas ramas dentro de esa rama, y asi sucesivamente.
# 
# La 3ra Dimension, dada por la variable dependiente, es la que nos sirve para la prediccion. Se saca la
# media  de la variable dependiente de cada conjunto. Una vez que ingrese un nuevo dato, se enontrara
# el nodo hoja al que pertenece haciendo uso de la informacion de las variables independientes, y se le
# asignara como variable dependiente, la media de ese conjunto para esa variable.
# 
# =======================================================================================================

################################################
###          IMPORTAR EL DATA SET            ###
################################################

setwd("~/Documentos/Udemy/machinelearning-az/datasets/Part 2 - Regression/Section 8 - Decision Tree Regression")
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[, 2:3]

#################################################################################
### Dividir el data set en conjunto de entrenamiento y conjunto de testing    ###
#################################################################################

# En este caso no se hara debido a la escases de datos.

# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Purchased, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# testing_set = subset(dataset, split == FALSE)


################################################
#            Escalado de variables             #
################################################

# Para los arboles de decision no suele hacerse escalado de variables, porque
# el algoritmo no utiliza distancias euclidianas.

# training_set[,2:3] = scale(training_set[,2:3])
# testing_set[,2:3] = scale(testing_set[,2:3])


########################################################
#  Ajustar el modelo con el dataset de Entrenamiento   #
########################################################

# install.packages("rpart")
# Crgamos la libreria "rpart" (Recursive Partitioning and Regression Tree)
library(rpart)
# Hacemos que el minimo numero de elementos de cada nodo hoja sea 1.
regression = rpart(formula = Salary ~ .,
                   data = dataset,
                   control = rpart.control(minsplit = 1))

################################################
#                PREDICCION                    #
################################################

y_pred = predict(regression, newdata = data.frame(Level = 6.5))

################################################
#        VISUALIZACION DE RESULTADOS           #
################################################

# Podria ser que obtengamos una linea horizontal (donde todos tuvieran el mismo sueldo),
# lo que podria estar pasando es que el arbol tuviera restricciones muy estrictas a la hora de dividir
# una rama en nodos hoja; como el tener un numero minimo de candidatos para cada nodo.
# Tambien el no. de divisiones, pues al arbol evita el overfitting, evitando nodos muy pequeños
# CHECAR PARAMETROS DE rpart()

library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level , y = dataset$Salary),
             color = "red") +
  geom_line(aes(x = dataset$Level, y = predict(regression, 
                                        newdata = data.frame(Level = dataset$Level))),
            color = "blue") +
  ggtitle("Predicción con Árbol de Decisión (Modelo de Regresión)") +
  xlab("Nivel del empleado") +
  ylab("Sueldo (en $)")

# En este plot podemos ver realmente lo que pasa en el algoritmo, al asignar la media como valor
# de la variable independiente. Ademas esto es lo que pasa cuando cada nodo contiene a 1 solo individuo.

x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() +
  geom_point(aes(x = dataset$Level , y = dataset$Salary),
             color = "red") +
  geom_line(aes(x = x_grid, y = predict(regression, 
                                        newdata = data.frame(Level = x_grid))),
            color = "blue") +
  ggtitle("Predicción con Árbol de Decisión (Modelo de Regresión)") +
  xlab("Nivel del empleado") +
  ylab("Sueldo (en $)")

File: datasets/Part 2 - Regression/Section 8 - Decision Tree Regression/decission_tree_regression.py
#!/home/kevinml/anaconda3/bin/python3.7
# -*- coding: utf-8 -*-
"""
Created on Thu Mar  7 19:04:40 2019

@author: juangabriel and Kevin Meza
"""

# CART (Classification and Regression Trees)
# =======================================================================================================
# 
#  Si graficamos en 2D datos de 2 variables independientes (X1,X2) que permitiran predecir  una variable
# dependiente, el algoritmo dividira los conjuntos de puntos con lineas rectas. El algoritmo mira la entropia,
# que tan juntos o dispersos pueden estar esos puntos o que similitudes tienen entre si. 
# Cada uno de estos conjuntos corresponde a un nodo hoja. Se continua haciendo divisiones hasta cierto
# punto, p.e. Cuando un nodo hoja se quede por lo menos al 5% de datos original.
# 
# La idea del algoritmo es que la cantidad de informacion aumenta (es mas acertada) cuando dividimos los 
# puntos en conjuntos o alguna otra regla, para hacer que el algoritmo converja y no haya Overffiting.
# 
# Al mismo tiempo que se divide el conjunto, se "genera" un arbol. si la primera division fuera que X1 < 20
# se generarian entonces 2 ramas. Luego si la sig. div. es que X2 < 170, pero solo en quellos con X1 > 20, 
# se generan 2 nuevas ramas dentro de esa rama, y asi sucesivamente.
# 
# La 3ra Dimension, dada por la variable dependiente, es la que nos sirve para la prediccion. Se saca la
# media  de la variable dependiente de cada conjunto. Una vez que ingrese un nuevo dato, se enontrara
# el nodo hoja al que pertenece haciendo uso de la informacion de las variables independientes, y se le
# asignara como variable dependiente, la media de ese conjunto para esa variable.
# 
# =======================================================================================================


# Regresión con Árboles de Decisión

# Cómo importar las librerías
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

################################################
###          IMPORTAR EL DATA SET            ###
################################################

dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values


#################################################################################
### Dividir el data set en conjunto de entrenamiento y conjunto de testing    ###
#################################################################################

# En este caso no se hara por la escases de datos

"""
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)
"""

################################################
#            Escalado de variables             #
################################################

# Para los arboles de decision no suele hacerse escalado de variables, porque
# el algoritmo no utiliza distancias euclidianas.
# Recomendado: Ver como sale el modelo sin escalar variables y luego escalar.

"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)"""

########################################################
#  Ajustar el modelo con el dataset de Entrenamiento   #
########################################################

#importamos la libreria para hcaer nuestro regresor
# https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeRegressor.html
from sklearn.tree import DecisionTreeRegressor
# Es recomendable checar los  parametros, como donde hacer cortes en el arbol.
# se suele utilizar la medida del error cuadrado medio, para minimizar la 
# diferencia al cuadrado entre la prediccion y el resultado. De esta forma
# se busca la forma de cortar minimiza los cuadrados de los errores.
# Otro parametro importante es el tipo de lineas que se trazan,
# verticales y horizontales o incluyendo diagonales, dependiento si se toman
# en cuenta 1 o 2 caracteristicas ara hacer las divisiones de los conjuntos.
# Mas parametros = no. max. de nodos hoja, no. max. de elementos por nodo
regression = DecisionTreeRegressor(random_state=0)
regression.fit(X, y)

################################################
#                PREDICCION                    #
################################################

y_pred = regression.predict([[6.5]])

################################################
#        VISUALIZACION DE RESULTADOS           #
################################################

plt.scatter(X, y, color="red")
plt.plot(X, regression.predict(X), color="blue")
plt.title("Modelo de Regresión")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo (en $)")
plt.show()

# En este plot podemos ver realmente lo que pasa en el algoritmo, al asignar la media como valor
# de la variable independiente. Ademas esto es lo que pasa cuando cada nodo contiene a 1 solo individuo.
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape(len(X_grid), 1)
plt.scatter(X, y, color="red")
plt.plot(X, regression.predict(X_grid), color="blue")
plt.title("Modelo de Regresión")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo (en $)")
plt.show()